### PR TITLE
Adloox Analytics Adapter: add new analytics adapter

### PR DESCRIPTION
**Pull request by jimdigriz**
## Type of change

- [X] New [Analytics adapter](https://github.com/corememltd/Prebid.js/blob/feature/adloox/modules/adlooxAnalyticsAdapter.js) ([documentation](https://github.com/corememltd/Prebid.js/blob/feature/adloox/modules/adlooxAnalyticsAdapter.md))
  - [x] New [integration example](https://github.com/corememltd/Prebid.js/blob/feature/adloox/integrationExamples/gpt/adloox.html)

## Description of change

Introduction of Adloox's analytics module.

* contact email of the adapter’s maintainer: contact@adloox.com and alex+adloox@coremem.com
* this is an official submission

Disclosures:

 * on `BID_WON` adds the Adloox post-buy verification script (~25kB gzipped from `j.adlooxtracking.com`) and locates it immediately after the ad slot in the DOM

This module automate the complex tasks an AdOps teams must perform when integrating with Adloox, including the script loading listed in the disclosure above. If the any of the scripts fail to load, it impacts only Adloox's measurement and does not affect the operation of Prebid.js or inventory fill for the publisher.

## Other information

 * provides a command processing queue to the other Adloox modules (two follow up PRs for the [AdServer video](https://github.com/prebid/Prebid.js/pull/6309) and [RTD](https://github.com/prebid/Prebid.js/pull/6310) modules) to provide configuration sharing and macro/URL expansion facilities
   * exists to reduce the burden on AdOps teams by preventing typos between each modules configuration that should be keep identical (eg. `{client,tag,platform}id`, `id[0-9]` parameters, ...)
   * created as `events.js` does not support custom events and is a shared event bus
   * works by exporting the same internal command array between all the modules; does not use the global namespace
     * these intra-module imports do trip the the eslint's 'prebid/validate-imports' check and has been [disabled for those modules](https://github.com/prebid/Prebid.js/pull/6174#issuecomment-758582001)

This PR was created on request in https://github.com/prebid/Prebid.js/pull/6174#issuecomment-763964406

Changes since that version:
 * word smithing

### CODE REVIEW

**Comment by jimdigriz** [comment]
I have no idea why CI is failing, and CircleCI will not let me see any details without permitting it to run wild with read/write access to all my repos, including the private ones :-/

Unit and the level of coverage testing Works For Me(tm).

**Comment by robertrmartinez** [comment]
@jimdigriz Sorry for the delay!

I will get to reviewing this today.

I have kicked off the Tests again, will let you know if it is just due to flaky-ness, or something actually wrong!

**Comment by jimdigriz** [comment]
> @jimdigriz Sorry for the delay!
>
> I will get to reviewing this today.

Not a problem, thank you for putting aside some of your time to do this.
 
> I have kicked off the Tests again, will let you know if it is just due to flaky-ness, or something actually wrong!

Yes, as mentioned in https://github.com/prebid/Prebid.js/pull/6310#issuecomment-783509982 (another PR) it Works-For-Me(tm)

Running node v14.15.5 in case that is a datapoint for you, though I really suspect that is not the cause of any potential race...

Let me know if you need anything or if I can help

**Comment by robertrmartinez** [comment]
@jimdigriz I kicked off the tests in that other PR you mentioned, let's see how it goes.

For this PR, it seems there is some issue with the `should preload the script on AUCTION_END only once` test when running in IE

I will try to look into it, but maybe there is some code using ES6 which is causing the issue.

Usually always the case for CircleCi IE failures.

**Comment by jimdigriz** [comment]
> I will try to look into it, but maybe there is some code using ES6 which is causing the issue.

Oh, my bad, I assumed there would be a transpiler in the mix to handle things like this.

I'll go through and get those bumped down to ES5 syntax.

**Comment by jimdigriz** [comment]
I tried all the unit tests in the Docker container (`circleci/node 12.16.1`) used in your CircleCI setup and the tests pass again there with all the ES6 syntax left intact having setup the environment with:
```
docker run -it --rm circleci/node:12.16.1 /bin/bash
cd ~
git clone https://github.com/corememltd/Prebid.js.git
cd Prebid.js
git checkout feature/adloox
npm install
sudo npm install -g gulp-cli
apt-get install --no-install-recommends chromium chromium-driver
```

```
circleci@e46a1cee3ff0:~/Prebid.js$ git log -n 1
commit 04c12f33ca1b11200a6b8b7db7f413e6fe6d31bc
Author: Alexander Clouter <alex@coremem.com>
Date:   Sat Feb 13 16:37:26 2021 +0000

    Adloox real time data module

circleci@e46a1cee3ff0:~/Prebid.js$ env CHROME_BIN=/usr/bin/chromium gulp test --nolintfix
[09:23:38] Using gulpfile ~/Prebid.js/gulpfile.js
[09:23:38] Starting 'test'...
[09:23:38] Starting 'clean'...
[09:23:38] Finished 'clean' after 15 ms
[09:23:38] Starting 'lint'...
[09:24:06] Finished 'lint' after 28 s
[09:24:06] Starting 'test'...

START:
⚠ ｢wdm｣: 
ℹ ｢wdm｣: Compiled with warnings.
ℹ ｢wdm｣: Compiling...
Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
⚠ ｢wdm｣: 
ℹ ｢wdm｣: Compiled with warnings.
24 02 2021 09:24:53.578:INFO [karma-server]: Karma v4.4.1 server started at http://0.0.0.0:9876/
24 02 2021 09:24:53.579:INFO [launcher]: Launching browsers ChromeCustom with concurrency 5
24 02 2021 09:24:53.785:INFO [launcher]: Starting browser ChromeHeadless
24 02 2021 09:24:54.550:INFO [HeadlessChrome 73.0.3683 (Linux 0.0.0)]: Connected on socket LPQ-SG-Jpss5CaStAAAA with id 16847991

Finished in 8.691 secs / 5.441 secs @ 09:25:05 GMT+0000 (Coordinated Universal Time)

SUMMARY:
✔ 7587 tests completed
ℹ 5 tests skipped
[09:25:05] Finished 'test' after 59 s
[09:25:05] Finished 'test' after 1.45 min
```

**Comment by robertrmartinez** [comment]
@jimdigriz Sorry for the confusion, we have a transpiler for the ES6 stuff, but some of the things like array.find, Object.entries, etc are not supported, we do not do polyfills.

I just always say ES6, when in reality it is a subset.

I'll figure out why these are failing

**Comment by smenzer** [comment]
@robertrmartinez any updates from your side?

File: modules/adlooxAnalyticsAdapter.js
Comment (robertrmartinez) [inline modules/adlooxAnalyticsAdapter.js:209]: @jimdigriz 

Part of the "policy" of analytics adapters has been "Just use the src/ajax.js lib to signal your event data"

As well as, we do not really allow inserting scripts / elements.

After discussing with the "Rules Committee" we decided that we will allow these kinds of things, since Pubs usually only pick one or two analytics adapters and usually do their due diligence in researching the things they do.

Part of this exception is we have now split out the Analytics Modules in the Prebid Docs repo to their own separate files.

So the ask here is for you to submit a corresponding doc and detail / not some of the different things your adapter does!

Here is the PR which broke out Analytics Adapters: https://github.com/prebid/prebid.github.io/pull/2727

Comment (jimdigriz) [reply modules/adlooxAnalyticsAdapter.js:209]: Submitted in https://github.com/prebid/Prebid.js/pull/6308#discussion_r589585261

Let me know if I have missed anything else blocking this from moving forward. Thanks.

**Comment by jimdigriz** [comment]
Anything I need to do here to get this merged?

Slightly concerned as it has now been a month (in this PR along) and there have been no comments about the code its-self or what may need changing. If something is needed here it would be great to hear about it sooner rather than later.

Let me know if I can help at all. Thanks

File: modules/adlooxAnalyticsAdapter.js
Comment (patmmccann) [reply modules/adlooxAnalyticsAdapter.js:209]: @jimdigriz could you duplicate all the prominent disclosures in your MD files back to this docs pr?

**Comment by patmmccann** [comment]
You're failing your own test though

**Comment by jimdigriz** [comment]
> @robertrmartinez any updates from your side?

Not on my system I am not.

There is a (stalled) dicussion about this in https://github.com/prebid/Prebid.js/pull/6308#issuecomment-785486587

I replicated your CircleCI container here (same version) and it also worked. I have no idea why your end is failing.

**Comment by harpere** [comment]
@jimdigriz you ran tests locally with IE11?

**Comment by jimdigriz** [comment]
> @jimdigriz you ran tests locally with IE11?

Nope, the contribution instructions did not ask me to https://github.com/prebid/Prebid.js#unit-testing-with-karma

I thought the transmogrification part of the build I thought would remove any IE11 problems with using ES6?

It will cost me ~$200 to get a one month subscription to Browserstack to run your unit tests on IE11, is this the only option available to me?

**Comment by robertrmartinez** [comment]
@jimdigriz 

That is why our main circle ci runs these for all PRs.

So what we need to do is find out why this same test fails everytime on IE 11

'should preload the script on AUCTION_END only once'

https://github.com/prebid/Prebid.js/pull/6308/files?file-filters%5B%5D=.js#diff-c15473e2f99d67a6fa50a35f9ded45937d126169fd71dd1b7cd0c7807b2d2270R137

specifically this line:
```
expect(insertElementStub.calledWith(sinon.match(isLinkPreloadAsScript))).to.true;
```

So would like to see if it is even calling the `insertElementStub`, and if so what is it called with.

Because it is not calling it exactly as excpeted.

File: modules/adlooxAnalyticsAdapter.js
Comment (jimdigriz) [reply modules/adlooxAnalyticsAdapter.js:209]: done

**Comment by robertrmartinez** [comment]
image

Did your last commit stomp over this one?

Wanted to see if this fixed it. but looks like it no longer is in the _spec

**Comment by jimdigriz** [comment]
> image
> 
> Did your last commit stomp over this one?

Yes, it was intended though.
 
> Wanted to see if this fixed it. but looks like it no longer is in the _spec

Whilst I was waiting for CircleCI it occurred to me that others know about this awfulness so I looked in your `utils.js` and it already has this workaround. https://github.com/prebid/Prebid.js/blob/70a41a51ac945b90c47224c294e64fc35742d604/src/utils.js#L1177

**Comment by robertrmartinez** [comment]
ahh yeah got ya.

Something is def going on in the bit of your code run by AUCTION_END event in IE.

I'll try to figure it out. I may have my own browserstack access to run locally.

**Comment by jimdigriz** [comment]
I looked under the sofa and found Adloox already has a forgotten about Browserstack subscription...now spelunking to see if there is an undocumented way to just test only my spec files rather than everything.

```
$ gulp e2e-test --host=test.localhost --file 'test/spec/modules/adloox{AnalyticsAdapter,AdServerVideo,RtdProvider}_spec.js'
[snipped]
[21:28:51] Appending pbjs.processQueue();
[21:28:51] Generating bundle: prebid.js
[21:28:51] Finished 'bound gulpBundle' after 249 ms
[21:28:51] Finished 'build-bundle-prod' after 26 s
[21:28:51] Starting 'injectFakeServerEndpoint'...
[21:28:51] Finished 'injectFakeServerEndpoint' after 270 ms
[21:28:51] Starting 'test'...
stdout: fake-server listening on http://localhost:4444


Execution of 3 spec files started at 2021-03-16T21:28:52.365Z

[1-0] RUNNING in chrome - /test/spec/modules/adlooxAnalyticsAdapter_spec.js
[2-0] RUNNING in firefox - /test/spec/modules/adlooxAnalyticsAdapter_spec.js
[0-0] RUNNING in ie - /test/spec/modules/adlooxAnalyticsAdapter_spec.js
[0-0]  Error:  Cannot use import statement outside a module
[0-0] FAILED in ie - /test/spec/modules/adlooxAnalyticsAdapter_spec.js
[1-0]  Error:  Cannot use import statement outside a module
[1-0] FAILED in chrome - /test/spec/modules/adlooxAnalyticsAdapter_spec.js
[2-0]  Error:  Cannot use import statement outside a module
[2-0] FAILED in firefox - /test/spec/modules/adlooxAnalyticsAdapter_spec.js

 "concise" Reporter:
========= Your concise report ==========
internet explorer (v11) on WINDOWS
All went well !!
========= Your concise report ==========
chrome (v80.0.3987.132) on WINDOWS
All went well !!
========= Your concise report ==========
firefox (v10.0) on windows
All went well !!


Spec Files:      0 passed, 3 failed, 3 total (100% completed) in 00:00:26 

[21:29:18] 'test' errored after 27 s
[21:29:18] Error: Tests failed with error: Error: Command failed: /home/alex/src/adloox/Prebid.js/node_modules/.bin/wdio /home/alex/src/adloox/Prebid.js/wdio.conf.js --spec test/spec/modules/adloox{AnalyticsAdapter,AdServerVideo,RtdProvider}_spec.js
    at /home/alex/src/adloox/Prebid.js/gulpfile.js:283:14
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
[21:29:18] 'e2e-test' errored after 54 s
```

**Comment by robertrmartinez** [comment]
@jimdigriz 

Not sure the --file command will work with multiple.

But you can just run the one file in this PR

`npm test -- --file=test/spec/modules/adlooxAnalyticsAdapter_spec.js `

**Comment by robertrmartinez** [comment]
Not sure that works when run through browserstack though.

**Comment by jimdigriz** [comment]
From my experience here, `file` works but only as a glob(ish) rule so have been using locally:
```
gulp test-coverage --file 'test/spec/modules/adloox{AnalyticsAdapter,AdServerVideo,RtdProvider}_spec.js' && gulp view-coverage
```

...but this passes on my system (Linux) as I use Firefox/Chrome. I do not have ready access to Windows/IE11.

**Comment by jimdigriz** [comment]
Magic undocumented incantation is:
```$ gulp test --browserstack --nolintfix --file test/spec/modules/adlooxAnalyticsAdapter_spec.js
[snipped]
16 03 2021 23:25:54.169:INFO [launcher]: Starting browser safari 12.0 (OS X Mojave) on BrowserStack
16 03 2021 23:25:54.186:ERROR [launcher.browserstack]: Can not start safari 12.0 (OS X Mojave)
  Error: getaddrinfo ENOTFOUND api.browserstack.com
16 03 2021 23:27:02.521:INFO [Edge 17.17134.0 (Windows 10.0.0)]: Connected on socket ZZGwOOs6Tqg9CaojAAAE with id 10253702
16 03 2021 23:27:51.416:INFO [Firefox 73.0.0 (Windows 10.0.0)]: Connected on socket comyJJuWg-bCAG0lAAAF with id 40162735
16 03 2021 23:28:31.400:INFO [Firefox 72.0.0 (Windows 10.0.0)]: Connected on socket _wF1J0e-JqvDmwwnAAAG with id 87378398
16 03 2021 23:29:11.384:INFO [Safari 13.1.2 (Mac OS X 10.15.6)]: Connected on socket H0idKn3Ik7tMtVoaAAAH with id 41186163

IE 11.0.0 (Windows 10.0.0): Executed 13 of 13 (1 FAILED) (0.101 secs / 0.027 secs)
Edge 18.18363.0 (Windows 10.0.0): Executed 13 of 13 SUCCESS (0.032 secs / 0.017 secs)
Chrome 80.0.3987 (Windows 10.0.0): Executed 13 of 13 SUCCESS (0.027 secs / 0.01 secs)
Chrome 79.0.3945 (Windows 10.0.0): Executed 13 of 13 SUCCESS (0.028 secs / 0.013 secs)
Edge 17.17134.0 (Windows 10.0.0): Executed 13 of 13 SUCCESS (0.054 secs / 0.02 secs)
Firefox 73.0.0 (Windows 10.0.0): Executed 13 of 13 SUCCESS (0.066 secs / 0.02 secs)
Firefox 72.0.0 (Windows 10.0.0): Executed 13 of 13 SUCCESS (0.06 secs / 0.017 secs)
Safari 13.1.2 (Mac OS X 10.15.6): Executed 13 of 13 SUCCESS (0.096 secs / 0.014 secs)
TOTAL: 1 FAILED, 103 SUCCESS


1) should preload the script on AUCTION_END only once
     Adloox Analytics Adapter process event
     expected false to be true
Error
   at AssertionError (webpack:///node_modules/assertion-error/index.js:74:1 <- test/spec/modules/adlooxAnalyticsAdapter_spec.js:8415:7)
   at Assertion.prototype.assert (webpack:///node_modules/chai/lib/chai/assertion.js:141:1 <- test/spec/modules/adlooxAnalyticsAdapter_spec.js:15483:7)
   at Anonymous function (webpack:///node_modules/chai/lib/chai/core/assertions.js:660:1 <- test/spec/modules/adlooxAnalyticsAdapter_spec.js:16173:5)
   at propertyGetter (webpack:///node_modules/chai/lib/chai/utils/addProperty.js:62:1 <- test/spec/modules/adlooxAnalyticsAdapter_spec.js:14540:9)
   at Anonymous function (webpack:///test/spec/modules/adlooxAnalyticsAdapter_spec.js:144:9 <- test/spec/modules/adlooxAnalyticsAdapter_spec.js:8803:9)


[23:29:14] 'test' errored after 5.43 min
[23:29:14] Error: Karma tests failed with exit code 1
    at /home/alex/src/adloox/Prebid.js/gulpfile.js:301:12
    at removeAllListeners (/home/alex/src/adloox/Prebid.js/node_modules/karma/lib/server.js:370:9)
    at /home/alex/src/adloox/Prebid.js/node_modules/karma/lib/server.js:381:13
    at Server.close (net.js:1606:9)
    at Object.onceWrapper (events.js:421:28)
    at Server.emit (events.js:327:22)
    at Server.EventEmitter.emit (domain.js:529:15)
    at emitCloseNT (net.js:1659:8)
    at processTicksAndRejections (internal/process/task_queues.js:79:21)
[23:29:14] 'test' errored after 5.8 min
```

Investigating...

**Comment by jimdigriz** [comment]
...and if you edit `gulpfile.js` to remove `.map(helpers.toCapitalCase)` from `browserOverride`, you can target just IE11 with: `gulp test --browserstack --browsers=bs_ie_11_windows_10 --nolint --nolintfix --file test/spec/modules/adlooxAnalyticsAdapter_spec.js`

https://github.com/prebid/Prebid.js/blob/70a41a51ac945b90c47224c294e64fc35742d604/gulpfile.js#L289

**Comment by jimdigriz** [comment]
Thank you IE...you can see it in the DOM and it leaks through in the tests, the `.hostname` includes the port.
```
17 03 2021 00:06:33.764:INFO [IE 11.0.0 (Windows 10.0.0)]: Connected on socket pXEaQ2tIS6tOezErAAAA with id 20640148
LOG LOG: 'https://j.adlooxtracking.com/ads/js/tfav_adl_127.js'
LOG LOG: 'https://j.adlooxtracking.com:443/ads/js/tfav_adl_127.js'
```

Passing `arg.href` also through `utils.parseUrl` then comparing the normalised .href of both works fine; I also have to move to using `.getAttribute` otherwise the `as="script"` is undefined.

**Comment by jimdigriz** [comment]
No idea why the tests are now failing:
```
alex@aineko:~/src/adloox/Prebid.js$ gulp test --browserstack
[snipped]
Edge 17.17134.0 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (21.318 secs / 12.468 secs)
Chrome 80.0.3987 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (15.375 secs / 8.5 secs)
Edge 18.18363.0 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (21.061 secs / 13.301 secs)
IE 11.0.0 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (33.04 secs / 22.535 secs)
Chrome 79.0.3945 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (18.801 secs / 9.553 secs)
Firefox 73.0.0 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (17.358 secs / 9.543 secs)
Firefox 72.0.0 (Windows 10.0.0): Executed 7655 of 7664 (skipped 9) SUCCESS (19.098 secs / 10.736 secs)
Safari 13.1.2 (Mac OS X 10.15.6): Executed 7655 of 7664 (skipped 9) SUCCESS (14.601 secs / 9.969 secs)
Safari 12.1.2 (Mac OS X 10.14.6): Executed 7655 of 0 (skipped 9) SUCCESS (42.822 secs / 12.088 secs)
TOTAL: 68895 SUCCESS

[14:44:31] Finished 'test' after 7.18 min
[14:44:31] Finished 'test' after 7.53 min
```

I am unable to log into CircleCI and see what it is complaining about as it demands I had over read/write access to all my repos (including my private ones).

**Comment by robertrmartinez** [comment]
image

This test is failing on every browser.

**Comment by robertrmartinez** [comment]
Sometimes it is flakey, I kicked off the test run again.

**Comment by jimdigriz** [comment]
Since yesterday, I did notice `adloader.js` and decided to use it to help others in the project audit where external JS loads come from. `test/mocks/adloaderStub.js` would hobble the test, but I do not understand why it would kick in for CircleCI tests but not when I run `gulp test --browserstack --nolintfix`

**Comment by jimdigriz** [comment]
Can someone tell me how to run the CircleCI tests and get the *same* results as you do here? It would let me get visibility and enable me to fix these problems myself.

I have inspected `.circleci/config.yml` and extracted out something that seems to replicate what CircleCI does but I get different results and the runs have a smell of non-determinism about them; I see failures on the non-Adloox modules.

Maybe someone could update the README to explain how testing actually is done (as `gulp test ...` with BrowserStack is *not* sufficient and does not match what CircleCI does) so it gives those outside the core project team a chance to contribute and not waste our time?

File: test/spec/modules/adlooxAnalyticsAdapter_spec.js
Comment (robertrmartinez) [inline test/spec/modules/adlooxAnalyticsAdapter_spec.js:155]: Should this test now be updated to just stub and validate the `loadExternalScript` function is called appropriately?

Now that the bid won handler is utilizing it?

File: test/mocks/adloaderStub.js
Comment (robertrmartinez) [inline test/mocks/adloaderStub.js:18]: So I am trying to pinpoint the issue, made my own branch with your changes https://github.com/prebid/Prebid.js/pull/6446

This change specifically seems to be causing problems, and I am banging my head on my desk trying to find out why.

All these tests running together, calling the adLoader, and some manipulating the script variable returned may be causing something to happen.

But when this above snippet is not checked in, the only errors are the Bid Won test in adlook spec:

image


But when this is checked in, it is always:

image

Comment (robertrmartinez) [reply test/mocks/adloaderStub.js:18]: Amazing error message, I know

Comment (ChrisHuie) [reply test/mocks/adloaderStub.js:18]: What is the context for modifying this function? Not sure I 100% understand why the adloaderStub is being updated.

Comment (jimdigriz) [reply test/mocks/adloaderStub.js:18]: For me it [seems impossible to get local testing to match CircleCI's output](https://github.com/prebid/Prebid.js/pull/6308#issuecomment-802009747) and so my only hope at the moment is stare at the code, make a change, submit it, wait 15->30mins for 'yes'/no' feedback and no details as the CircleCI output is completely unavailable to me (unless I give CircleCI full read/write access to all my public/private repos).

...it is desperation at this point, an alternative methodology would be welcomed hugely.

Comment (ChrisHuie) [reply test/mocks/adloaderStub.js:18]: Ah. Sorry. I didn't realize it was just for testing purposes.

Comment (patmmccann) [reply test/mocks/adloaderStub.js:18]: this is the test that fails on edge, right?   https://github.com/prebid/Prebid.js/blob/595fc0a08b5f97a319b81f8099394623218efe7c/test/spec/modules/userId_spec.js#L2190

Comment (patmmccann) [reply test/mocks/adloaderStub.js:18]: > (unless I give CircleCI full read/write access to all my public/private repos).

or do the pr from a different github login ?

Comment (jimdigriz) [reply test/mocks/adloaderStub.js:18]: > > (unless I give CircleCI full read/write access to all my public/private repos).
> 
> or do the pr from a different github login ?

...or use a CI provider that does not force all contributors to do this? We could argue both ways, it does not help :)

The problem here is that the described testing in the project readme does not match the testing you actually use and blocks PR contributions. So even if I handed everything over to CircleCI you still are asking me to wait ~30 minutes just to get any feedback and nothing actionable other than "Script error" to work with.

What would be infinitely helpful is if someone describes *exactly* what to type/run/use to get matching results. I have access to Browserstack, I run the same commands I see in `.circleci/config.yml` but I get different results. This needs to be fixed either by sending me clue on what I am doing wrong or updating your "how to test your changes" documentation.

Lets start with, can you show me what you run on your own workstation that replicates CircleCI's output? With that, I can then figure out where I am going wrong or it may highlight deficiencies in the documentation, either way it seems to be the right next step.

Comment (ChrisHuie) [reply test/mocks/adloaderStub.js:18]: I have tried a few things here and it seems to be a race condition on the behavioral testing side with the stub and different analytics adapters touching the DOM.

Comment (jimdigriz) [reply test/mocks/adloaderStub.js:18]: I am happy to soak time into this if someone can explain the process to get matching results on my local workstation to what is coughed up by CircleCI.

**Comment by robertrmartinez** [comment]
@jimdigriz 

I have pushed a fix to me test branch.

If this looks okay to you, maybe you can adopt this into yours?

https://github.com/prebid/Prebid.js/pull/6446

Basically, I instead of relying on the script inserted, I am just checking the loadExternalScript stub to make sure it was called with the right params. please see [this line](https://github.com/prebid/Prebid.js/pull/6446/files#diff-c15473e2f99d67a6fa50a35f9ded45937d126169fd71dd1b7cd0c7807b2d2270R176-R181) specifically

Also note I had to make changes to `test/mocks/adloaderStub.js` to export the `loadExternalScript` stub so we can inspect it

**Comment by jimdigriz** [comment]
> @jimdigriz
> 
> I have pushed a fix to my test branch.
> 
> If this looks okay to you, maybe you can adopt this into yours?
> 
> #6446

This works for me. Thank you for putting your own time into solving this and helping out, your efforts are hugely appreciated.

Now the CI testing works, what else do I need to do to get this PR into shape?

**Comment by robertrmartinez** [comment]
@jimdigriz No problem.

Sorry it was such a pain the the rear, for most of us this is just a hobby / not our full time job, so there's a lot about the code base which has undocumented quirks. I want to hash out some time for myself to make this all a bit easier in the near future.

I will give your PR one more good look here after I eat breakfast, and we will get it merged before the next Prebid.js release next week!

Thanks for your patience.

File: modules/adlooxAnalyticsAdapter.js
Comment (robertrmartinez) [inline modules/adlooxAnalyticsAdapter.js:210]: Should this be using the `loadExternalScript` function as well?

Comment (jimdigriz) [reply modules/adlooxAnalyticsAdapter.js:210]: No, we only want to load the post-buy verification JS at the `BID_WON` event but as we use the same URL for all the slots under auction we take the opportunity to prefetch the JS so it loads as close to the event as possible.

**Review by robertrmartinez — APPROVED** [review]

**Review by ChrisHuie — APPROVED** [review]

**Comment by jimdigriz** [comment]
Thank you to everyone involved in helping get this done, much obliged.

As a quick query, is there another project I need to submit a PR to to get us listed as an option to enable on https://docs.prebid.org/download.html or does this trickle in and get updated by some other independent process?

**Pull request merged**
